### PR TITLE
Add utility functions to talk to AWS.

### PR DESCRIPTION
**Pull request by kdelee**
The functions in integrade.tests.aws_utils are meant to facilitate generating usage data on AWS, and provide a consistent way of referring to different AWS accounts and the images associated with them.

Updates #28

@elyezer @ironfroggy would appreciate some input on what you think of this not just from correctness perspective but also usability. 

Proposed configuration to use these functions:
Use `~/.aws/config` to allow boto to switch between different "profiles" -- this is actually configuration for the `aws` cli client.
Use `~/.config/integrade/aws_config` to allow users to define the images associated with a aws profile (named in the `~/.aws/config` file). This way we could associate data with the image and not have to use the ami to get instances started (can share image names and perhaps profile names so we can have different accounts and AMIs but everyone be able to run the same test cases as long as we have "customer1" and it has a rhel image defined under "rhel1".

Example `~/.config/integrade/aws_config`
```
profiles:
  customer1:
      role_arn: "arn:aws:iam::1234567890:role/role-for-cloudigrade"
      images:
        rhel1:
            is_rhel: True
            image_id: 'ami-543'
            is_private_shared: False

        centos1:
            is_rhel: False
            image_id: 'ami-123'
            is_private_shared: False
```

### CODE REVIEW

File: integrade/tests/aws_utils.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+    :param aws_profile: (string) Name of profile as defined in config file
+    :param image_name: (string) Name of image as defined in config file
+    :param count: (int) Number of instances to create.
+    :param run_time: (double) Desired time to let instances run (in seconds)
+
+    :returns: (double) Time from when instances were all live until all
+         instances were terminated (in seconds).
+    """
+    response = create_instances(aws_profile, image_name, count=count)
```
Comment (ironfroggy) [inline integrade/tests/aws_utils.py]: Let's move the `wait_until_running()` pool inside `create_instance()` itself. That will make using it a lot cleaner and, later, if we use it in other helpers, we'll keep its usage consistently correct.

Comment (kdelee) [reply integrade/tests/aws_utils.py]: :+1:

File: tests/test_config.py [hunk 2]
```diff
@@ -4,9 +4,26 @@
 
 import pytest
 
+import xdg
+
+import yaml
```
[the remark this answers is not in the document]
Comment (kdelee) [reply tests/test_config.py]: yes! good catch

**Comment by kdelee** [comment]
@elyezer could you do a final pass over?

The test I added helped me work out a few more kinks about what info I would need, and I realized I needed to associate the ARN with the images.

I also was able to reproduce https://github.com/cloudigrade/cloudigrade/issues/357 with my new test when I used a private shared image....

I honestly have not got that test to pass yet, even when while watching the celery logs and it looks like everything is OK. I think that the test is correct, and either it 1) takes longer than 10 min for houndigrade to do its job :fearful: or 2) https://github.com/cloudigrade/cloudigrade/pull/350 did not quite get us to where we need to be from the API side to see the inspection state get updated.

Basically, I think the test is correct.

File: integrade/tests/api/v1/test_instances.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    aws_cfg = config.get_aws_config()
+    aws_profile = 'customer1'
+    # create some instances to detect on creation
+    created_instance_ids = aws_utils.create_instances(
+        aws_profile, 'rhel1', count=3)
+    # add new instances to list of instances to terminate after test
+    instances_to_terminate.extend(zip([aws_profile] * 3, created_instance_ids))
+    cloud_account = {
+        'account_arn': aws_cfg['profiles'][aws_profile]['role_arn'],
+        'resourcetype': 'AwsAccount'
```
Comment (elyezer) [inline integrade/tests/api/v1/test_instances.py]: Something tells me that AwsAccount could be a constant

File: integrade/tests/api/v1/test_instances.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    }
+    client.post(
+        urls.CLOUD_ACCOUNT,
+        payload=cloud_account,
+        auth=auth
+    )
+    list_instances = client.get(urls.INSTANCE, auth=auth).json()
```
Comment (elyezer) [inline integrade/tests/api/v1/test_instances.py]: It seems that you could have used the json_handler for this api client

File: integrade/tests/aws_utils.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    """Test if aws config is missing and return boolean."""
+    try:
+        config.get_aws_config()
+        return False
+    except ConfigFileNotFoundError:
+        return True
+
+
+# apply as decorator on top of tests that need AWS config to be present
+aws_config_needed = minversion = pytest.mark.skipif(
```
Comment (elyezer) [inline integrade/tests/aws_utils.py]: I don't see minversion used anywhere, am I missing something?

File: integrade/tests/aws_utils.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    This is useful when you want to make sure there are no running instances
+    in an account. Terminated instances eventually disappear from the list of
+    instances visible in the EC2 console or via describe_instances(), but this
+    is cannot be controlled by the user (happens on the AWS backend).
+    """
+    client = aws_session(aws_profile).client('ec2')
+    instances_to_terminate = []
+    for reservation in client.describe_instances().get('Reservations', []):
+        for instance in reservation.get('Instances'):
+            if instance['State']['Code'] != 48:
```
Comment (elyezer) [inline integrade/tests/aws_utils.py]: Maybe worth adding a comment stating which is the state code name or create a constant to make easy to read.

File: integrade/tests/conftest.py [hunk 7]
```diff
@@ -35,3 +38,25 @@ def drop_account_data():
...
+    """
+    instances_to_terminate = []
+
+    yield instances_to_terminate
+
+    if instances_to_terminate:
+        num_instances = len(instances_to_terminate)
+        with Pool(num_instances) as p:
+            p.map(
+                terminate_instance, instances_to_terminate)
```
Comment (elyezer) [inline integrade/tests/conftest.py]: Maybe join could be used whenever a Pool was created to make sure the all processes complete before moving on.

https://docs.python.org/3.5/library/multiprocessing.html#multiprocessing.pool.Pool.join

Comment (kdelee) [reply integrade/tests/conftest.py]: Reading https://docs.python.org/2/library/multiprocessing.html#multiprocessing.pool.multiprocessing.Pool.map I understood "blocks until it finishes" to mean no additional join was needed

File: integrade/tests/aws_utils.py (hunk 5 above)
Comment (kdelee) [reply integrade/tests/aws_utils.py]: woops copy pasta

File: integrade/tests/conftest.py (hunk 7 above)
Comment (elyezer) [reply integrade/tests/conftest.py]: I've been thinking a little bit more about using the close and join and I checked the Pool's source code and realized that with the context manager it will terminate (see https://github.com/python/cpython/blob/6f0eb93183519024cb360162bdd81b9faec97ba6/Lib/multiprocessing/pool.py#L574-L575).

That said and considering that the `map` will only return after completing all tasks calling close and join is useless as you pointed. I saw that you've added the close and join so feel free to remove it. I apologize for that, when I was reviewing I didn't had the fact that you were using Pool as a context manager.
